homeassistant/components/energy/sensor.py

```diff
@@ -383,27 +362,58 @@ def _update_cost(self) -> None:
         old_energy_value = float(self._last_energy_sensor_state.state)
         cur_value = cast(float, self._attr_native_value)
 
-        if energy_price_unit is None:
-            converted_energy_price = energy_price
-        else:
-            converter: Callable[[float, str, str], float]
-            if energy_unit in VALID_ENERGY_UNITS:
-                converter = unit_conversion.EnergyConverter.convert
-            else:
-                converter = unit_conversion.VolumeConverter.convert
-
-            converted_energy_price = converter(
-                energy_price,
-                energy_unit,
-                energy_price_unit,
-            )
+        converted_energy_price = self._convert_energy_price(
+            energy_price, energy_price_unit, energy_unit
+        )
 
         self._attr_native_value = (
             cur_value + (energy - old_energy_value) * converted_energy_price
         )
 
         self._last_energy_sensor_state = energy_state
 
+    def _get_energy_price(
+        self,
+        valid_units: set[str],
+        default_unit: str | None,
+    ) -> tuple[float | None, str | None] | None:
+        if self._config["entity_energy_price"] is None:
+            return cast(float, self._config["number_energy_price"]), default_unit
+
+        energy_price_state = self.hass.states.get(self._config["entity_energy_price"])
+        if energy_price_state is None:
+            return None
+
+        try:
+            energy_price = float(energy_price_state.state)
+        except ValueError:
+            return (None, None)
```

Comment:
No a big fan of this tuple, but we do need to distinguish the entity-not-found case from this one.
Moving the call to `self._reset(energy_state)` to this method would work, but I prefer to keep the method without side-effects, as those would not be obvious from the caller.

Reply:
I agree this is ugly. Why don't you just let the error bubble up, and let the caller handle it?